### ISSUE TITLE
Failed advertising of topics

### ISSUE THREAD

**Issue by lorenzoriano** [ros-drivers/rosserial#63]
The following code snippet works fine:

``` c++
    ros::NodeHandle  nh;
    nh.initNode(rosSrvrIp);

    std_msgs::Float32 steer_msg;
    ros::Publisher pub("/steering_controller/steer", &steer_msg);
    nh.advertise(pub);

    while (!nh.connected())
        nh.spinOnce();

    while (1) {
        steer_msg.data = 5;
        pub.publish(&steer_msg);
        nh.spinOnce();
        sleep(1);
    }
```

However if I run the spinOnce cycles after having advertised the topic, then rosserial_python fails with: 

```
Tried to publish before configured, topic id 125
```

This is the code snipped that causes the failure:

``` c++
    ros::NodeHandle  nh;
    nh.initNode(rosSrvrIp);

    while (!nh.connected())
        nh.spinOnce();

    std_msgs::Float32 steer_msg;
    ros::Publisher pub("/steering_controller/steer", &steer_msg);
    nh.advertise(pub);

    while (1) {
        steer_msg.data = 5;
        pub.publish(&steer_msg);
        nh.spinOnce();
        sleep(1);
    }
```

Is there s fix for this problem? I'm not sure if it's caused by the client libraries or by rosserial_python.

**Comment by mikepurvis** [comment]
The issue is the advertise() function only caches information about the topic being advertised—the actual exchange occurs during the connection handshake. By the time you're already connected, it's too late to advertise more topics.

This is a limitation of rosserial_client which is confusing and not well-documented. It could be lifted, of course, but at a cost of increased complexity in the client.

**Comment by lorenzoriano** [comment]
What would it take to change the client? Will it affect the overall system? This bug might prevent people from using rosserial effectively, especially since it's not documented.

**Comment by mikepurvis** [comment]
Closing as wontfix.

**Comment by Nick-Zhang1996** [comment]
When does this handshake take place? How do we ensure advertise() is called before the actual exchange?

**Comment by romainreignier** [comment]
@Nick-Zhang1996 I am not too familiar with the implementation details of `rosserial`, but it seems that the handshake takes place with the `negotiateTopics()` method which is called in the `spinOnce()` method: https://github.com/ros-drivers/rosserial/blob/fe1e233f39d56931b5e223c5e53a2d9c2f9fb664/rosserial_client/src/ros_lib/ros/node_handle.h#L324

So assuming that you call `spinOnce()` only after the initialization (after `initNode()`, `advertise()` and `subscribe()`), you should be OK.
